File: api/src/main/java/com/example/sms/domain/model/master/product/Product.java
package com.example.sms.domain.model.master.product;

import com.example.sms.domain.type.ProductType;
import com.example.sms.domain.type.money.Money;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Value;

import java.util.List;

/**
 * 商品
 */
@Value
@AllArgsConstructor
@NoArgsConstructor(force = true)
public class Product {
    ProductCode productCode; // 商品コード
    ProductName productName; // 商品名
    ProductType productType; // 商品区分
    Money sellingPrice; // 販売単価
    Money purchasePrice; // 仕入単価
    Money costOfSales; // 売上原価
    Integer taxType; // 税区分
    ProductCategoryCode productCategoryCode; // 商品分類コード
    Integer miscellaneousType; // 雑区分
    Integer stockManagementTargetType; // 在庫管理対象区分
    Integer stockAllocationType; // 在庫引当区分
    SupplierCode supplierCode; // 仕入先コード
    List<SubstituteProduct> substituteProduct; // 代替商品
    List<Bom> boms; // 部品表
    List<CustomerSpecificSellingPrice> customerSpecificSellingPrices; // 顧客別販売単価

    public static Product of(String productCode, String productFormalName, String productAbbreviation, String productNameKana, String productType, Integer sellingPrice, Integer purchasePrice, Integer costOfSales, Integer taxType, String productClassificationCode, Integer miscellaneousCategory, Integer stockManagementTargetCategory, Integer stockAllocationCategory, String supplierCode, Integer supplierBranchNumber) {
        ProductName productName = ProductName.of(productFormalName, productAbbreviation, productNameKana);
        return new Product(ProductCode.of(productCode), productName, ProductType.fromCode(productType), Money.of(sellingPrice), Money.of(purchasePrice), Money.of(costOfSales), taxType, ProductCategoryCode.of(productClassificationCode), miscellaneousCategory, stockManagementTargetCategory, stockAllocationCategory, SupplierCode.of(supplierCode, supplierBranchNumber), List.of(), List.of(), List.of());
    }

    public static Product of(Product product, List<SubstituteProduct> substituteProduct, List<Bom> boms, List<CustomerSpecificSellingPrice> customerSpecificSellingPrices) {
        ProductName productName = ProductName.of(product.productName.getProductFormalName(), product.productName.getProductAbbreviation(), product.productName.getProductNameKana());
        return new Product(product.productCode, productName, product.productType, product.sellingPrice, product.purchasePrice, product.costOfSales, product.taxType, product.productCategoryCode, product.miscellaneousType, product.stockManagementTargetType, product.stockAllocationType, product.supplierCode, substituteProduct, boms, customerSpecificSellingPrices);
    }
}

File: app/src/models/product.ts
type ProductCodeType = {
    value: string;
    businessType: string;
    itemType: string;
    livestockType: string;
    serialNumber: number;
};

type ProductNameType = {
    productFormalName: string;
    productAbbreviation: string;
    productNameKana: string;
};

type PriceType = {
    amount: number;
    currency: string;
};

type QuantityType = {
    amount: number;
    unit: string;
}

type SupplierCodeType = {
    value: string;
    branchNumber: number;
};

export type SubstituteProductType = {
    productCode: ProductCodeType;
    substituteProductCode: ProductCodeType;
    priority: number;
};

export type BomType = {
    productCode: ProductCodeType;
    componentCode: ProductCodeType;
    componentQuantity: QuantityType;
}

export type CustomerSpecificSellingPriceType = {
    productCode: ProductCodeType;
    customerCode: string;
    sellingPrice: PriceType;
};

export type ProductType = {
    productCode: ProductCodeType;
    productName: ProductNameType;
    productType: string;
    sellingPrice: PriceType;
    purchasePrice: PriceType;
    costOfSales: PriceType;
    taxType: string;
    productCategoryCode: { value: string };
    miscellaneousType: string;
    stockManagementTargetType: string;
    stockAllocationType: string;
    supplierCode: SupplierCodeType;
    substituteProduct: SubstituteProductType[];
    boms: BomType[];
    customerSpecificSellingPrices: CustomerSpecificSellingPriceType[];
    checked: boolean;
    addFlag: boolean;
    deleteFlag: boolean;
};

export type ProductResourceType = {
    productCode: string;
    productFormalName: string;
    productAbbreviation: string;
    productNameKana: string;
    productType: string;
    sellingPrice: number;
    purchasePrice?: number;
    costOfSales: number;
    taxType: string;
    productClassificationCode?: string;
    miscellaneousType?: string;
    stockManagementTargetType?: string;
    stockAllocationType?: string;
    supplierCode: string;
    supplierBranchNumber?: number;
    substituteProduct?: SubstituteProductType[];
    boms?: BomType[];
    customerSpecificSellingPrices?: CustomerSpecificSellingPriceType[];
    addFlag: boolean;
    deleteFlag: boolean;
}

type ProductCategoryCode = {
    value: string;
};

export type ProductCategoryType = {
    productCategoryCode: ProductCategoryCode;
    productCategoryName: string;
    productCategoryHierarchy: number;
    productCategoryPath: string;
    lowestLevelDivision: number;
    products: ProductType[];
    checked: boolean;
};

export type ProductCategoryResourceType = {
    productCategoryCode: string;
    productCategoryName?: string;
    productCategoryHierarchy: number;
    productCategoryPath?: string;
    lowestLevelDivision?: number;
    products?: ProductResourceType[];
};

export const mapToProductResource = (product: ProductType): ProductResourceType => {
    return {
        productCode: product.productCode.value,
        productFormalName: product.productName.productFormalName,
        productAbbreviation: product.productName.productAbbreviation,
        productNameKana: product.productName.productNameKana,
        productType: product.productType,
        sellingPrice: product.sellingPrice.amount,
        purchasePrice: product.purchasePrice.amount,
        costOfSales: product.costOfSales.amount,
        taxType: product.taxType,
        productClassificationCode: product.productCategoryCode.value,
        miscellaneousType: product.miscellaneousType,
        stockManagementTargetType: product.stockManagementTargetType,
        stockAllocationType: product.stockAllocationType,
        supplierCode: product.supplierCode.value,
        supplierBranchNumber: product.supplierCode.branchNumber,
        substituteProduct: product.substituteProduct,
        boms: product.boms,
        customerSpecificSellingPrices: product.customerSpecificSellingPrices,
        addFlag: product.addFlag,
        deleteFlag: product.deleteFlag
    };
}

export const mapToProductCategoryResource = (productCategory: ProductCategoryType): ProductCategoryResourceType => {
    return {
        productCategoryCode: productCategory.productCategoryCode.value,
        productCategoryName: productCategory.productCategoryName,
        productCategoryHierarchy: productCategory.productCategoryHierarchy,
        productCategoryPath: productCategory.productCategoryPath,
        lowestLevelDivision: productCategory.lowestLevelDivision,
        products: productCategory.products.map(product => mapToProductResource(product))
    };
}
